directorywatcher/directorywatcher.go

```diff
@@ -0,0 +1,288 @@
+// file in order to refresh it's stored data.
+type DirectoryWatcher interface {
+	// Get returns the latest, parsed data from the DirectoryWatcher.
+	Get() interface{}
+
+	// Stop stops the DirectoryWatcher.
+	//
+	// After Stop is called you won't get any updates on the file content,
+	// but you can still call Get to get the last content before stopping.
+	//
+	// It's OK to call Stop multiple times.
+	// Calls after the first one are essentially no-op.
+	Stop()
+}
+
+// AddFile is a type of function that should be ran to handle
+// adding data after a file has been parsed
+type AddFile func(d interface{}, file interface{}) (data interface{}, err error)
+
+// RemoveFile is a type of function that should be ran to handle removing
+// data after a file has been removed from the watcher
+type RemoveFile func(d interface{}, path string) (data interface{}, err error)
+
+// Result is the return type of New. Use Get function to get the actual data.
+type Result struct {
+	data atomic.Value
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+// Config defines the config to be used in New function.
+//
+// Can be deserialized from YAML.
+type Config struct {
+	// The path to the file to be watched, required.
+	Path string `yaml:"path"`
+
+	// The parser to parse the data load, required.
+	Parser filewatcher.Parser
+
+	Adder AddFile
+
+	Remover RemoveFile
+
+	// Optional. When non-nil, it will be used to log errors,
+	// either returned by parser or by the underlying file system watcher.
+	// Please note that this does not include errors returned by the first parser
+	// call, which will be returned directly.
+	Logger log.Wrapper `yaml:"logger"`
+
+	// Optional. When <=0 DefaultMaxFileSize will be used instead.
+	//
+	// This is the soft limit,
+	// we will also auto add a hard limit which is 10x (see HardLimitMultiplier)
+	// of soft limit.
+	//
+	// If the soft limit is violated,
+	// the violation will be reported via log.ErrorWithSentry,
+	// but it does not stop the normal parsing process.
+	//
+	// If the hard limit is violated,
+	// The loading of the file will fail immediately.
+	MaxFileSize int64 `yaml:"maxFileSize"`
+}
+
+// Get returns the latest parsed data from the file watcher.
+//
+// Although the type is interface{},
+// it's guaranteed to be whatever actual type is implemented inside Parser.
+func (r *Result) Get() interface{} {
+	return r.data.Load()
+}
+
+// Stop stops the file watcher.
+//
+// After Stop is called you won't get any updates on the file content,
+// but you can still call Get to get the last content before stopping.
+//
+// It's OK to call Stop multiple times.
+// Calls after the first one are essentially no-op.
+func (r *Result) Stop() {
+	r.cancel()
+}
+
+func (r *Result) watcherLoop(
+	watcher *fsnotify.Watcher,
+	parser filewatcher.Parser,
+	add AddFile,
+	remove RemoveFile,
+	softLimit, hardLimit int64,
+	logger log.Wrapper,
+) {
+	for {
+		select {
```

Comment:
In my opinion, trying to deal with the specific operations is too complex to be justified.

I would recommend instead using a debounce.

Whenever you get any notification, start a timer.  Say, 100ms, 500ms, or 1s depending on how fast you need to react.  If you get another notification before it fires, reset the timer.  When the timer fires, do a complete pass over all files and load them.

This way you can write one set of logic that handles the files the way they appear on the filesystem, you don't need to worry about delta-encoding and the _many_ issues inherent therein.

Reply:
I do like this but I still think extra logic for merging data between files will be needed somehow